src/webhook/cached-verifier.ts

```diff
@@ -0,0 +1,62 @@
+import nacl from "tweetnacl";
+import { logger } from "../logger.js";
+
+export class CachedVerifier {
+  private expirationTimes: { [message: string]: number } = {};
+
+  constructor(public knownPublicKeys: string[]) {}
+
+  public verify(signature: string): boolean {
+    if (!this.expirationTimes[signature]) {
+      this.removedExpired();
+      return this.verifyMessage(signature);
+    }
+
+    if (this.now() >= this.expirationTimes[signature]) {
+      this.removedExpired();
+      return false;
+    }
+
+    return true;
+  }
+
+  private verifyMessage(message: string) {
+    try {
+      const msg = Buffer.from(message, "base64url");
+      const signature = msg.subarray(0, nacl.sign.signatureLength);
+      const payloadBuffer = msg.subarray(nacl.sign.signatureLength);
+      const payload = JSON.parse(payloadBuffer.toString("utf-8"));
+
+      if (this.now() >= payload.exp) {
```

Comment:
We shouldn't use `payload.exp`

We should include `signature-expiry` in header